File: src/server/crdt/shopping/shopping_item.go
package crdt

import (
	crdt "sdle-server/crdt/generic"
	g01 "sdle-server/proto"
)

type ShoppingItem struct {
	replicaID  string
	dotContext *crdt.DotContext
	itemID     string
	name       *crdt.MVReg[string]
	quantity   *crdt.CCounter
	acquired   *crdt.CCounter
	deleted    *crdt.DWFlag
}

func NewShoppingItem(replicaID string, itemID string) *ShoppingItem {
	dotContext := crdt.NewDotContext()

	name := crdt.NewMVReg[string](replicaID)
	name.SetContext(dotContext)

	deletedFlag := crdt.NewDWFlag(replicaID)
	deletedFlag.SetContext(dotContext)

	quantity := crdt.NewCCounter(replicaID)
	quantity.SetContext(dotContext)

	acquired := crdt.NewCCounter(replicaID)
	acquired.SetContext(dotContext)

	return &ShoppingItem{
		replicaID:  replicaID,
		dotContext: dotContext,
		itemID:     itemID,
		name:       name,
		deleted:    deletedFlag,
		quantity:   quantity,
		acquired:   acquired,
	}
}

func (si *ShoppingItem) ReplicaID() string {
	return si.replicaID
}

func (si *ShoppingItem) Name() string {
	// Assumes a single value in the MVReg
	values := si.name.Read()
	if len(values) > 0 {
		return values[0]
	}
	return ""
}

func (si *ShoppingItem) SetName(name string) *ShoppingItem {
	delta := NewShoppingItem(si.replicaID, si.itemID)

	delta.name = si.name.Write(name)
	delta.SetContext(delta.name.Context())

	return delta
}

func (si *ShoppingItem) ItemID() string {
	return si.itemID
}

func (si *ShoppingItem) SetItemID(itemID string) {
	si.itemID = itemID
}

func (si *ShoppingItem) Quantity() uint64 {
	return uint64(si.quantity.Read())
}

func (si *ShoppingItem) IncQuantity(amount int64) *ShoppingItem {
	delta := NewShoppingItem(si.replicaID, si.itemID)

	// Prevent negative quantity values
	amount = max(-si.quantity.Read(), amount)

	quantityDelta := si.quantity.Inc(amount)
	delta.quantity = quantityDelta
	delta.SetContext(quantityDelta.Context())

	return delta
}

func (si *ShoppingItem) Acquired() uint64 {
	return uint64(si.acquired.Read())
}

func (si *ShoppingItem) IncAcquired(amount int64) *ShoppingItem {
	delta := NewShoppingItem(si.replicaID, si.itemID)
	currQuantity := int64(si.quantity.Read())
	currAcquired := int64(si.acquired.Read())

	// Prevent negative acquired values
	amount = max(-currAcquired, amount)

	// Prevent acquired from exceeding quantity
	amount = min(currQuantity-currAcquired, amount)

	acquiredDelta := si.acquired.Inc(amount)
	delta.acquired = acquiredDelta
	delta.SetContext(acquiredDelta.Context())

	return delta
}

func (si *ShoppingItem) Deleted() bool {
	return si.deleted.Read()
}

func (si *ShoppingItem) Delete() *ShoppingItem {
	delta := NewShoppingItem(si.replicaID, si.itemID)

	nonEnabledDelta := si.deleted.Enable()
	delta.deleted = nonEnabledDelta
	delta.SetContext(nonEnabledDelta.Context())

	return delta
}

func (si *ShoppingItem) Restore() *ShoppingItem {
	delta := NewShoppingItem(si.replicaID, si.itemID)

	nonEnabledDelta := si.deleted.Disable()
	delta.deleted = nonEnabledDelta
	delta.SetContext(nonEnabledDelta.Context())

	return delta
}

func (si *ShoppingItem) Context() *crdt.DotContext {
	return si.dotContext
}

func (si *ShoppingItem) SetContext(ctx *crdt.DotContext) {
	si.dotContext = ctx

	si.name.SetContext(ctx)
	si.quantity.SetContext(ctx)
	si.acquired.SetContext(ctx)
	si.deleted.SetContext(ctx)
}

func (si *ShoppingItem) Reset() *ShoppingItem {
	delta := NewShoppingItem(si.replicaID, si.itemID)

	delta.name = si.name.Reset()
	delta.quantity = si.quantity.Reset()
	delta.acquired = si.acquired.Reset()
	delta.deleted = si.deleted.Enable()

	delta.dotContext.Join(delta.name.Context())
	delta.dotContext.Join(delta.quantity.Context())
	delta.dotContext.Join(delta.acquired.Context())
	delta.dotContext.Join(delta.deleted.Context())

	return delta
}

func (si *ShoppingItem) IsNull() bool {
	return si.name.IsNull() && si.quantity.IsNull() && si.acquired.IsNull() && si.deleted.IsNull()
}

func (si *ShoppingItem) Join(other *ShoppingItem) {
	// Save original context to restore after merging quantity
	originalContext := si.dotContext.Clone()

	// Merge itemID if it is empty
	// This is needed when empty items are created automatically by ORMap
	if si.itemID == "" {
		si.itemID = other.itemID
	}

	si.name.Join(other.name)
	si.dotContext.Copy(originalContext)

	si.quantity.Join(other.quantity)
	si.dotContext.Copy(originalContext)

	si.acquired.Join(other.acquired)
	si.dotContext.Copy(originalContext)

	si.deleted.Join(other.deleted)
	// No need to restore original context here

	si.dotContext.Join(other.dotContext)
}

func (si *ShoppingItem) Clone() *ShoppingItem {
	clone := NewShoppingItem(si.replicaID, si.itemID)

	clone.dotContext = si.dotContext.Clone()
	clone.name = si.name.Clone()
	clone.quantity = si.quantity.Clone()
	clone.acquired = si.acquired.Clone()
	clone.deleted = si.deleted.Clone()

	clone.name.SetContext(clone.dotContext)
	clone.quantity.SetContext(clone.dotContext)
	clone.acquired.SetContext(clone.dotContext)
	clone.deleted.SetContext(clone.dotContext)

	return clone
}

func (si ShoppingItem) String() string {
	return "ShoppingItem{" +
		"replicaID: " + si.replicaID +
		", itemID: " + si.itemID +
		", name: " + si.name.String() +
		", deleted: " + si.deleted.String() +
		", quantity: " + si.quantity.String() +
		", acquired: " + si.acquired.String() +
		"}"
}

func (si *ShoppingItem) ToProto() *g01.ShoppingItem {
	return &g01.ShoppingItem{
		Name:     ((*crdt.StringMVReg)(si.name)).ToProto(),
		Quantity: si.quantity.ToProto(),
		Acquired: si.acquired.ToProto(),
		Deleted:  si.deleted.ToProto(),
	}
}

func ShoppingItemFromProto(protoItem *g01.ShoppingItem, replicaID string, itemID string, ctx *crdt.DotContext) *ShoppingItem {
	name := (*crdt.MVReg[string])(crdt.StringMVRegFromProto(protoItem.GetName(), replicaID, ctx))
	quantity := crdt.CCounterFromProto(protoItem.GetQuantity(), replicaID, ctx)
	acquired := crdt.CCounterFromProto(protoItem.GetAcquired(), replicaID, ctx)
	deleted := crdt.DWFlagFromProto(protoItem.GetDeleted(), replicaID, ctx)

	return &ShoppingItem{
		replicaID:  replicaID,
		dotContext: ctx,
		itemID:     itemID,
		name:       name,
		quantity:   quantity,
		acquired:   acquired,
		deleted:    deleted,
	}
}

File: src/client/types/crdt/shopping/shopping-item.ts
import CCounter from "../generic/ccounter";
import DotContext from "../generic/dot-context";
import { ShoppingItem as ShoppingItemProto, CCounter as CCounterProto, StringMVReg, DWFlag as DWFlagProto } from "@/lib/proto/global";
import DWFlag from "../generic/dwflag";
import MVReg from "../generic/mvreg";

export default class ShoppingItem {
    private replicaId: string;
    private itemId: string;
    private name: MVReg<string>;
    private quantity: CCounter;
    private acquired: CCounter;
    private deleted: DWFlag;
    private dotContext: DotContext;

    constructor(replicaId: string, itemId: string) {
        this.replicaId = replicaId;
        this.itemId = itemId;
        this.name = new MVReg<string>(replicaId);
        this.quantity = new CCounter(replicaId);
        this.acquired = new CCounter(replicaId);
        this.deleted = new DWFlag(replicaId);
        this.dotContext = new DotContext();

        this.quantity.setContext(this.dotContext);
        this.acquired.setContext(this.dotContext);
        this.deleted.setContext(this.dotContext);
        this.name.setContext(this.dotContext);
    }

    public getReplicaId(): string {
        return this.replicaId;
    }

    public getItemId(): string {
        return this.itemId;
    }

    public setItemId(itemId: string): void {
        this.itemId = itemId;
    }

    public getName(): string {
        // Assume single value for name
        return this.name.read().at(0) || "";
    }

    public setName(name: string): ShoppingItem {
        const delta = new ShoppingItem(this.replicaId, this.itemId);

        delta.name = this.name.write(name);
        delta.setContext(delta.name.getContext());

        return delta;
    }

    public getContext(): DotContext {
        return this.dotContext;
    }

    public setContext(context: DotContext): void {
        this.dotContext = context;
        this.quantity.setContext(context);
        this.acquired.setContext(context);
        this.deleted.setContext(context);
        this.name.setContext(context);
    }

    public getQuantity(): number {
        return this.quantity.read();
    }

    public incQuantity(amount: number): ShoppingItem {
        const delta = new ShoppingItem(this.replicaId, this.itemId);

        // Prevent negative quantity values
        amount = Math.max(-this.quantity.read(), amount);

        const quantityDelta = this.quantity.inc(amount);
        delta.quantity = quantityDelta;
        delta.setContext(quantityDelta.getContext());

        return delta;
    }

    public getAcquired(): number {
        return this.acquired.read();
    }

    public incAcquired(amount: number): ShoppingItem {
        const delta = new ShoppingItem(this.replicaId, this.itemId);

        const currQuantity = this.quantity.read();
        const currAcquired = this.acquired.read();

        // Prevent negative acquired values
        amount = Math.max(-currAcquired, amount);

        // Prevent acquired from exceeding quantity
        amount = Math.min(currQuantity - currAcquired, amount);

        const acquiredDelta = this.acquired.inc(amount);
        delta.acquired = acquiredDelta;
        delta.setContext(acquiredDelta.getContext());

        return delta;
    }

    public isDeleted(): boolean {
        return this.deleted.read();
    }

    public delete(): ShoppingItem {
        const delta = new ShoppingItem(this.replicaId, this.itemId);

        const deletedDelta = this.deleted.enable();
        delta.deleted = deletedDelta;
        delta.setContext(deletedDelta.getContext());

        return delta;
    }

    public restore(): ShoppingItem {
        const delta = new ShoppingItem(this.replicaId, this.itemId);

        const deletedDelta = this.deleted.disable();
        delta.deleted = deletedDelta;
        delta.setContext(deletedDelta.getContext());

        return delta;
    }

    public reset(): ShoppingItem {
        const delta = new ShoppingItem(this.replicaId, this.itemId);

        delta.name = this.name.reset();
        delta.quantity = this.quantity.reset();
        delta.acquired = this.acquired.reset();
        delta.deleted = this.deleted.enable();

        delta.getContext().join(delta.name.getContext());
        delta.getContext().join(delta.quantity.getContext());
        delta.getContext().join(delta.acquired.getContext());
        delta.getContext().join(delta.deleted.getContext());
        delta.setContext(delta.getContext());  // Set the context for new children

        return delta;
    }

    public isNull(): boolean {
        return this.name.isNull() && this.quantity.isNull() && this.acquired.isNull() && this.deleted.isNull();
    }

    public join(other: ShoppingItem): void {
        const originalCtx = this.dotContext.clone();

        // Merge itemID if it are empty
        // This is needed when empty items are created automatically by ORMap
        if (this.itemId === "") {
            this.itemId = other.itemId;
        }

        this.name.join(other.name);
        this.dotContext.copy(originalCtx);

        this.quantity.join(other.quantity);
        this.dotContext.copy(originalCtx);

        this.acquired.join(other.acquired);
        this.dotContext.copy(originalCtx);

        this.deleted.join(other.deleted);
        // No need to restore context here

        this.dotContext.join(other.getContext());
    }

    public clone(): ShoppingItem {
        const clone = new ShoppingItem(this.replicaId, this.itemId);

        clone.dotContext = this.dotContext.clone();
        clone.name = this.name.clone();
        clone.quantity = this.quantity.clone();
        clone.acquired = this.acquired.clone();
        clone.deleted = this.deleted.clone();

        clone.name.setContext(clone.dotContext);
        clone.quantity.setContext(clone.dotContext);
        clone.acquired.setContext(clone.dotContext);
        clone.deleted.setContext(clone.dotContext);

        return clone;
    }

    public toProto(): ShoppingItemProto {
        return ShoppingItemProto.create({
            name: this.name.toStringProto(),
            deleted: this.deleted.toProto(),
            quantity: this.quantity.toProto(),
            acquired: this.acquired.toProto(),
        });
    }

    public static fromProto(proto: ShoppingItemProto, replicaId: string, itemId: string, ctx: DotContext): ShoppingItem {
        const item = new ShoppingItem(replicaId, itemId);

        item.setContext(ctx);
        item.name = MVReg.fromStringProto(proto.name as StringMVReg, replicaId, ctx);
        item.quantity = CCounter.fromProto(proto.quantity as CCounterProto, replicaId, ctx);
        item.acquired = CCounter.fromProto(proto.acquired as CCounterProto, replicaId, ctx);
        item.deleted = DWFlag.fromProto(proto.deleted as DWFlagProto, replicaId);

        return item;
    }
}